### PR TITLE
Adding MQTT module(client)

### PR DESCRIPTION
**Pull request by DanielBallaSZTE**
This patch adds the MQTT client support to IoT.js.

Co-authored-by: Daniel Vince <vinced@inf.u-szeged.hu>
IoT.js-DCO-1.0-Signed-off-by: Daniel Balla dballa@inf.u-szeged.hu

### CODE REVIEW

File: src/modules/iotjs_module_mqtt.c [hunk 1]
```diff
@@ -0,0 +1,584 @@
...
+         strlen(enc_remaining_length));
+  offset += strlen(enc_remaining_length);
+
+  // Variable header
+  memcpy(buffer_wrap->buffer + offset, variable_header_protocol,
+         sizeof(variable_header_protocol));
+  offset += sizeof(variable_header_protocol);
+
+  // Flags
+  memcpy(buffer_wrap->buffer + offset, &connect_flag, sizeof(connect_flag));
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.c]: No need `memcpy` here, simply `buffer_wrap->buffer[offset] = connect_flags`

File: src/modules/iotjs_module_mqtt.c [hunk 2]
```diff
@@ -0,0 +1,584 @@
...
+  jerry_value_t jclient_id =
+      iotjs_jval_get_property(joptions, IOTJS_MAGIC_STRING_CLIENTID);
+  jerry_value_t jusername =
+      iotjs_jval_get_property(joptions, IOTJS_MAGIC_STRING_USERNAME);
+  jerry_value_t jpassword =
+      iotjs_jval_get_property(joptions, IOTJS_MAGIC_STRING_PASSWORD);
+  jerry_value_t jkeepalive =
+      iotjs_jval_get_property(joptions, IOTJS_MAGIC_STRING_KEEPALIVE);
+
+  uint8_t connect_flag = 0;
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.c]: plural: flags

File: src/modules/iotjs_module_mqtt.c [hunk 3]
```diff
@@ -0,0 +1,584 @@
...
+  char *buffer = buffer_wrap->buffer;
+
+  char first_byte = buffer[0];
+  char remaining_length = buffer[1];
+  char packet_type = (first_byte >> 4) & 0b00001111;
+
+  switch (packet_type) {
+    case CONNACK:
+      if (remaining_length != 2) {
+        return JS_CREATE_ERROR(COMMON, "MQTT: CONNACK packet is corrupted");
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.c]: Maybe we should think about closing the connection.

File: src/modules/iotjs_module_mqtt.c [hunk 4]
```diff
@@ -0,0 +1,584 @@
...
+
+  jerry_value_t jsref = JS_GET_ARG(0, object);
+  jerry_value_t jparam = JS_GET_ARG(1, object);
+
+  iotjs_bufferwrap_t *buffer_wrap = iotjs_bufferwrap_from_jbuffer(jparam);
+  char *buffer = buffer_wrap->buffer;
+
+  char first_byte = buffer[0];
+  char remaining_length = buffer[1];
+  char packet_type = (first_byte >> 4) & 0b00001111;
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.c]: 0x0f

File: src/modules/iotjs_module_mqtt.h [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+    uint8_t retain : 1;
+    uint8_t qos : 2;
+    uint8_t dup : 1;
+    uint8_t type : 4;
+  } bits;
+} MQTTHeader;
+
+enum {
+  // Reserved bit, must be 0
+  FLAG_RESERVED = 1 << 0,
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.h]: MQTT_FLAG_....

File: src/modules/iotjs_module_mqtt.h [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+  FLAG_USERNAME = 1 << 7
+} iotjs_mqtt_connect_flag_t;
+
+
+typedef struct {
+  struct {
+    uint8_t msb;
+    uint8_t lsb;
+    char *buffer;
+  } data;
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.h]: Do we need this struct in struct?

Comment (DanielBallaSZTE) [reply src/modules/iotjs_module_mqtt.h]: No, we don't thanks for pointing out.

**Comment by DanielBallaSZTE** [comment]
@zherczeg I've changed those things, thank you for your time!

File: src/modules/iotjs_module_mqtt.c [hunk 7]
```diff
@@ -0,0 +1,677 @@
...
+    (*offset)++;
+   } while ((c & 128) != 0);
+
+  return remaining_length;
+}
+
+
+static void iotjs_mqtt_calculate_msb_lsb(uint16_t len, uint8_t *msb,
+                                         uint8_t *lsb) {
+  *msb = (uint8_t)((len & 0xFF00) >> 8);
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.c]: no need `& 0xFF00`

File: src/modules/iotjs_module_mqtt.c [hunk 8]
```diff
@@ -0,0 +1,677 @@
...
+  size_t full_len = sizeof(header_byte) + payload_len + variable_header_len +
+                    strlen(enc_remaining_length);
+
+  jerry_value_t jbuff = iotjs_bufferwrap_create_buffer(full_len);
+  iotjs_bufferwrap_t *buffer_wrap = iotjs_bufferwrap_from_jbuffer(jbuff);
+
+  uint8_t *buff_ptr = (uint8_t *)buffer_wrap->buffer;
+
+  // Fixed header
+  *buff_ptr++ = header_byte;
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.c]: I think this is the typical case when you don't need comment. It is obvious what you do here.

File: src/modules/iotjs_module_mqtt.c [hunk 9]
```diff
@@ -0,0 +1,677 @@
...
+      return JS_CREATE_ERROR(COMMON, "MQTT: Unknown error");
+  }
+}
+
+
+/*
+ * Waits for a uint32_t remaining_length.
+ * Returns the encoded value with the correct size.
+ */
+static char *iotjs_encode_remaining_length(uint32_t len) {
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.c]: I feel this is overcomplicated. You could write the data directly into the stream.

File: src/modules/iotjs_module_mqtt.c [hunk 10]
```diff
@@ -0,0 +1,677 @@
...
+  size_t rc = 0;
+  // It is an uint32_t at maximum
+  char temp_buff[4];
+  do {
+    char d = len % 128;
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.c]: `len & 0x7F;`

File: src/modules/iotjs_module_mqtt.c [hunk 11]
```diff
@@ -0,0 +1,677 @@
...
+    len /= 128;
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.c]: `len >>= 7`

File: src/modules/iotjs_module_mqtt.h [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+} iotjs_mqtt_connection_error;
+
+/*
+ * First byte of the message's fixed header.
+ * Contains:
+ * - MQTT Control Packet type,
+ * - Specific flags to each MQTT Control Packet.
+ */
+typedef struct {
+#ifdef MQTT_REVERSED
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.h]: How can this be reversed???? There is no bit order, only byte order.

Comment (DanielBallaSZTE) [reply src/modules/iotjs_module_mqtt.h]: This is left over from some old code, we never use it, it's going to get deleted when we clean up the header file.

File: src/modules/iotjs_module_mqtt.c (hunk 9 above)
Comment (DanielBallaSZTE) [reply src/modules/iotjs_module_mqtt.c]: We could, but we need the size of that data to allocate the buffer.

Comment (DanielBallaSZTE) [reply src/modules/iotjs_module_mqtt.c]: I figured out a solution for this.

**Comment by DanielBallaSZTE** [comment]
@zherczeg I have updated the PR, thank you!

File: src/modules/iotjs_module_mqtt.c [hunk 13]
```diff
@@ -0,0 +1,777 @@
...
+
+static uint32_t iotjs_decode_remaining_length(char *buffer, size_t *offset) {
+  unsigned char c;
+  uint32_t remaining_length = 0;
+  uint32_t length = 0;
+  uint32_t multiplier = 1;
+
+  do {
+    if (++length > IOTJS_MODULE_MQTT_MAX_REMAINING_LENGTH_BYTES) {
+      return JS_CREATE_ERROR(COMMON, "MQTT: The arrived message is too big");
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.c]: You should return with a special value to show error. E.g. `UINT32_MAX`, since the message size is limited to 256 Mbyte.

File: src/modules/iotjs_module_mqtt.c [hunk 14]
```diff
@@ -0,0 +1,777 @@
...
+    remaining_length += (c & 127) * multiplier;
+    multiplier *= 128;
+    (*offset)++;
+  } while ((c & 128) != 0);
+
+  return remaining_length;
+}
+
+
+static void iotjs_mqtt_calculate_msb_lsb(uint16_t len, uint8_t *msb,
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.c]: In general I would do the following for such functions:
```
static uint8_t* iotjs_mqtt_encode_length(uint8_t *buffer, uint16_t len)
{
  buffer[0] = (uint8_t)(len >> 8);
  buffer[1] = (uint8_t)(len & 0xff);
  return buffer + 2;
}
```
So they copy their arguments into the stream and advances the string pointer.

File: src/modules/iotjs_module_mqtt.c [hunk 15]
```diff
@@ -0,0 +1,777 @@
...
+
+  jerry_value_t jbuff = iotjs_bufferwrap_create_buffer(full_len);
+  iotjs_bufferwrap_t *buffer_wrap = iotjs_bufferwrap_from_jbuffer(jbuff);
+
+  uint8_t *buff_ptr = (uint8_t *)buffer_wrap->buffer;
+
+  *buff_ptr++ = header_byte;
+  *buff_ptr++ = remaining_length;
+  *buff_ptr++ = packet_identifier_msb;
+  *buff_ptr++ = packet_identifier_lsb;
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.c]: For the last bytes you can use `buff_ptr[0]` - `buff_ptr[3]`. The last ++ has no effect anyway.

File: src/modules/iotjs_module_mqtt.c [hunk 16]
```diff
@@ -0,0 +1,777 @@
...
+JS_FUNCTION(MqttUnsubscribe) {
+  DJS_CHECK_THIS();
+  DJS_CHECK_ARGS(1, object);
+
+  jerry_value_t jtopic = JS_GET_ARG(0, object);
+
+  iotjs_mqtt_payload_t topic_payload =
+      iotjs_create_payload(iotjs_bufferwrap_from_jbuffer(jtopic));
+
+  uint8_t header_byte = 0;
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.c]: `uint8_t header_byte = (UNSUBSCRIBE << 4);`
No need every step in separeted lines.

File: src/modules/iotjs_module_mqtt.c [hunk 17]
```diff
@@ -0,0 +1,777 @@
...
+
+  uint8_t packet_identifier_msb = 0;
+  uint8_t packet_identifier_lsb = 0;
+
+  size_t payload_len = topic_payload.size;
+  size_t variable_header_len =
+      sizeof(packet_identifier_msb) + sizeof(packet_identifier_lsb);
+  uint32_t remaining_length = payload_len + variable_header_len;
+  size_t full_len = sizeof(header_byte) + payload_len + variable_header_len +
+                    get_remaining_length_size(remaining_length);
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.c]: I would compute these in message order:
`sizeof(header_byte) + get_remaining_length_size(remaining_length) + variable_header_len + payload_len`

File: src/modules/iotjs_module_mqtt.c [hunk 18]
```diff
@@ -0,0 +1,777 @@
...
+
+  jerry_value_t jbuff = iotjs_bufferwrap_create_buffer(full_len);
+  iotjs_bufferwrap_t *buffer_wrap = iotjs_bufferwrap_from_jbuffer(jbuff);
+
+  uint8_t *buff_ptr = (uint8_t *)buffer_wrap->buffer;
+
+  *buff_ptr++ = header_byte;
+
+  buff_ptr += iotjs_encode_remaining_length(remaining_length, buff_ptr);
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.c]: if `iotjs_encode_remaining_length` returns with buffer, no need `+=`. The `=` is always faster since the compiler does not need to preserve the value of `buffer`.

File: src/modules/iotjs_module_mqtt.c [hunk 19]
```diff
@@ -0,0 +1,777 @@
...
+
+  jerry_value_t joptions = JS_GET_ARG(0, object);
+
+  jerry_value_t jtopic =
+      iotjs_jval_get_property(joptions, IOTJS_MAGIC_STRING_TOPIC);
+  jerry_value_t jqos =
+      iotjs_jval_get_property(joptions, IOTJS_MAGIC_STRING_QOS);
+
+  uint8_t qos = 0;
+  if (!jerry_value_is_undefined(jqos) && !jerry_value_has_error_flag(jqos)) {
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.c]: I think there is a utility now which converts strings and buffers to char*, which is exactly what you need here. It also returns with error if the conversion is not possible.

File: src/modules/iotjs_module_mqtt.c [hunk 20]
```diff
@@ -0,0 +1,777 @@
...
+  DJS_CHECK_ARGS(2, object, object);
+
+  jerry_value_t jsref = JS_GET_ARG(0, object);
+  jerry_value_t jparam = JS_GET_ARG(1, object);
+
+  iotjs_bufferwrap_t *buffer_wrap = iotjs_bufferwrap_from_jbuffer(jparam);
+  char *buffer = buffer_wrap->buffer;
+
+  char first_byte = buffer[0];
+  char packet_type = (first_byte >> 4) & 0b00001111;
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.c]: `0x0f`. Please don't use 0b

**Comment by DanielBallaSZTE** [comment]
@zherczeg Thank you again for the review, I have modified lots of things, and removed some unused things from the header file. What we still need are some checks if the user gives the correct input.
Besides from that, I'm open for any suggestions :)

File: src/js/mqtt.js [hunk 21]
```diff
@@ -0,0 +1,308 @@
...
+  this._protocol = options.protocol || 'net';
+
+  this._clientOptions = Object.create(options, {
+    host: { value: options.host || '127.0.0.1'},
+    port: { value: options.port || 8883 },
+    qos: { value: options.qos || 0 },
+    keepalive: { value: options.keepalive || 60 },
+  });
+
+  this._socket = new net.Socket();
```
Comment (zherczeg) [inline src/js/mqtt.js]: Is it a simple `net.Socket` even if tls is specified? I would remove the _protocol member and would simply accept a duplex stream as options.socket. If it is not specified, a new `net.Socket` is created.

File: src/js/mqtt.js [hunk 22]
```diff
@@ -0,0 +1,308 @@
...
+  this._socket.on('data', function(data) {
+    jsref.ondata(jsref, data);
+  });
+  this._socket.on('error', function(e) {
+    jsref.emit('error', e);
+  });
+  this._socket.on('end', this.ondisconnect);
+};
+
+MQTTClient.prototype.disconnect = function(error) {
```
Comment (zherczeg) [inline src/js/mqtt.js]: I think `end` is more common than `disconnect`.

File: src/js/mqtt.js [hunk 23]
```diff
@@ -0,0 +1,292 @@
...
+MQTTClient.prototype.unsubscribe = function(topic) {
+  if (!Buffer.isBuffer(topic)) {
+    topic = new Buffer(topic);
+  }
+
+  var buf = native.unsubscribe(topic);
+  this._socket.write(buf);
+}
+
+MQTTClient.prototype.onpingresp = function() {
```
[the remark this answers is not in the document]
Comment (zherczeg) [reply src/js/mqtt.js]: Let me rephrase the question. These seem internal functions. However they are exposed through `MQTTClient.prototype`. What about simply create a `function onpingresp() {...}`. This function would be not visible externally, but could use anything internally.

File: src/js/mqtt.js [hunk 24]
```diff
@@ -0,0 +1,308 @@
...
+}
+
+function getClient(connectionOptions) {
+  if (util.isUndefined(connectionOptions.clientId)) {
+    connectionOptions.clientId = defaultClientId();
+  }
+  if (!Buffer.isBuffer(connectionOptions.clientId)) {
+    connectionOptions.clientId = new Buffer(connectionOptions.clientId);
+  }
+  return MQTTClient(connectionOptions);
```
Comment (zherczeg) [inline src/js/mqtt.js]: `return new MQTTClient(connectionOptions);`

File: src/modules/iotjs_module_mqtt.h [hunk 25]
```diff
@@ -0,0 +1,130 @@
...
+  DISCONNECT = 0xE
+} iotjs_mqtt_control_packet_type;
+
+/*
+ * The values of the Quality of Service.
+ */
+enum {
+  QoS0 = 0b00, // At most once delivery.
+  QoS1 = 0b01, // At least once delivery.
+  QoS2 = 0b10  // Exactly once delivery.
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.h]: Please use 0x, or here you can use decimal numbers.

File: src/modules/iotjs_module_mqtt.c [hunk 26]
```diff
@@ -0,0 +1,719 @@
...
+    case UNAUTHORISED:
+      return JS_CREATE_ERROR(COMMON, "MQTT: Connection refused: unauthorised");
+    default:
+      return JS_CREATE_ERROR(COMMON, "MQTT: Unknown error");
+  }
+}
+
+
+static uint8_t *iotjs_encode_remaining_length(uint32_t len,
+                                              unsigned char *buffer) {
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.c]: `uint8_t *buffer`. Usually buffer (= destination byte-stream pointer) is always the first argument for such functions.

File: src/modules/iotjs_module_mqtt.c [hunk 27]
```diff
@@ -0,0 +1,719 @@
...
+  while (len != 0) {
+    len >>= 7;
+    n++;
+  }
+
+  return n;
+}
+
+
+static uint32_t iotjs_decode_remaining_length(char *buffer, size_t *offset) {
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.c]: Since `get_remaining_length_size` is called earlier, I would move the error checking there, and just use some asserts here. Perhaps get_remaining_length_size could return 0 on error, since at least 1` byte is required if the encoding is possible.

File: src/modules/iotjs_module_mqtt.c [hunk 28]
```diff
@@ -0,0 +1,719 @@
...
+  uint32_t length = 0;
+  uint32_t multiplier = 1;
+
+  do {
+    if (++length > IOTJS_MODULE_MQTT_MAX_REMAINING_LENGTH_BYTES) {
+      return UINT32_MAX;
+    }
+    c = (unsigned char)buffer[*offset];
+    remaining_length += (c & 127) * multiplier;
+    multiplier *= 128;
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.c]: `multiplier <<= 7`

Comment (zherczeg) [reply src/modules/iotjs_module_mqtt.c]: Btw usually shift is faster than multiplication, so you could simply use a shift:
```
remaining_length += (c & 0x7F) << shift;
shift += 7
```
Also `|=` can be used instead of `+=` here, but they are equally fast so they don't provide any advantage.

File: src/modules/iotjs_module_mqtt.c [hunk 29]
```diff
@@ -0,0 +1,719 @@
...
+    (*offset)++;
+  } while ((c & 128) != 0);
+
+  return remaining_length;
+}
+
+
+static void iotjs_mqtt_calculate_length(uint16_t *len, uint8_t msb,
+                                        uint8_t lsb) {
+  *len = (msb << 8) | lsb;
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.c]: What about `static uint16_t iotjs_mqtt_calculate_length(uint8_t msb, uint8_t lsb) { ... }`?

File: src/modules/iotjs_module_mqtt.c [hunk 30]
```diff
@@ -0,0 +1,719 @@
...
+}
+
+
+static uint8_t *iotjs_mqtt_string_serialize(uint8_t *dst_buffer,
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.c]: For me encode/serialize is equally ok, but please only use on of them. E.g `iotjs_encode_remaining_length` -> `iotjs_mqtt_serialize_remaining_length` or `iotjs_mqtt_string_encode`. Perhaps encode sounds a bit better, but either of them is ok.

File: src/modules/iotjs_module_mqtt.c [hunk 31]
```diff
@@ -0,0 +1,719 @@
...
+    connect_flags |= MQTT_FLAG_PASSWORD;
+  }
+  if (!jerry_value_is_undefined(jkeepalive)) {
+    uint16_t len = jerry_get_number_value(jkeepalive);
+    keep_alive_msb = (uint8_t)(len >> 8);
+    keep_alive_lsb = (uint8_t)(len & 0x00FF);
+  }
+
+
+  iotjs_bufferwrap_t *client_id = iotjs_bufferwrap_from_jbuffer(jclient_id);
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.c]: No type check here. I mentioned that there is a function which creates iotjs string from strings/buffers, or return NULL on error.

File: src/modules/iotjs_module_mqtt.c [hunk 32]
```diff
@@ -0,0 +1,719 @@
...
+  header_byte |= (jerry_get_boolean_value(jretain));
+
+  iotjs_bufferwrap_t *message_payload = iotjs_bufferwrap_from_jbuffer(jmessage);
+  iotjs_bufferwrap_t *topic_payload = iotjs_bufferwrap_from_jbuffer(jtopic);
+
+  uint8_t packet_identifier_lsb = 0;
+  uint8_t packet_identifier_msb = 0;
+
+  // IF NOT QOS
+  if (!jerry_value_is_undefined(jqos) &&
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.c]: There isa  to_boolean function, which returns with `bool` (never throws error). I would use it here.

File: src/modules/iotjs_module_mqtt.c [hunk 33]
```diff
@@ -0,0 +1,719 @@
...
+        return JS_CREATE_ERROR(COMMON, "MQTT: PingRESP packet is corrupted");
+      }
+      jerry_value_t fn =
+          iotjs_jval_get_property(jsref, IOTJS_MAGIC_STRING_ONPINGRESP);
+      iotjs_make_callback(fn, jsref, iotjs_jargs_get_empty());
+      jerry_release_value(fn);
+      break;
+    }
+    case DISCONNECT:
+      return JS_CREATE_ERROR(COMMON, "MQTT: Broker disconnected the client");
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.c]: Why this is an error? Isn't it a notification?

File: src/modules/iotjs_module_mqtt.c [hunk 34]
```diff
@@ -0,0 +1,719 @@
...
+
+    case CONNECT:
+    case SUBSCRIBE:
+    case UNSUBSCRIBE:
+    case PINGREQ:
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.c]: Server cannot send ping?

File: src/modules/iotjs_module_mqtt.c [hunk 35]
```diff
@@ -0,0 +1,719 @@
...
+      iotjs_make_callback(fn, jsref, iotjs_jargs_get_empty());
+      jerry_release_value(fn);
+      break;
+    }
+    case PUBLISH: {
+      MQTTHeader header = { 0 };
+      header.bits.type = PUBLISH;
+      header.bits.dup = first_byte & 0x08;
+      header.bits.qos = (first_byte & 0x06) >> 1;
+      header.bits.retain = first_byte & 0x01;
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.c]: I think more checks would be useful here. E.g. jerry offers utf8 validating functions.

File: src/modules/iotjs_module_mqtt.c [hunk 36]
```diff
@@ -0,0 +1,719 @@
...
+
+      iotjs_jargs_destroy(&args);
+      jerry_release_value(jmessage);
+      jerry_release_value(jtopic);
+
+      break;
+    }
+    case PUBACK: {
+      if (remaining_length != 2) {
+        return JS_CREATE_ERROR(COMMON, "MQTT: PUBACK packet is corrupted");
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.c]: Is the connection is aborted on error? I think the standard expects it.

File: src/modules/iotjs_module_mqtt.c [hunk 37]
```diff
@@ -0,0 +1,719 @@
...
+
+  jerry_release_value(jmessage);
+  jerry_release_value(jtopic);
+  jerry_release_value(jretain);
+
+  return jbuff;
+}
+
+
+JS_FUNCTION(MqttHandle) {
```
Comment (zherczeg) [inline src/modules/iotjs_module_mqtt.c]: What happens if the package is not fully arrived?

**Comment by zherczeg** [comment]
Travis: `Signed-off-by message is incorrect. The following line should be at the end of the e5135d6 commit's message: 'IoT.js-DCO-1.0-Signed-off-by: Daniel Vince vinced@inf.u-szeged.hu'. `

File: src/js/mqtt.js (hunk 22 above)
Comment (DanielBallaSZTE) [reply src/js/mqtt.js]: There actually is a packet type which has the name `DISCONNECT`, causing the server to properly disconnect the client.

File: src/js/mqtt.js (hunk 24 above)
Comment (DanielBallaSZTE) [reply src/js/mqtt.js]: Yeah about the prototype thing, we haven't refactored the JS code yet, were focusing on the C side. 
We were planning on changing them as you mentioned it.

File: src/modules/iotjs_module_mqtt.c (hunk 31 above)
Comment (DanielBallaSZTE) [reply src/modules/iotjs_module_mqtt.c]: I don't think it is needed since JS side a buffer is always created for the client id, and that's passed to the C side.

File: src/modules/iotjs_module_mqtt.c (hunk 37 above)
Comment (vincedani) [reply src/modules/iotjs_module_mqtt.c]: The maximum TCP package size is 64K, but due to memory limitations the IoT.js can't handle such a big packages.

**Comment by DanielBallaSZTE** [comment]
@daeyeon I've updated the commit! Thank you

**Comment by yichoi** [comment]
LGTM
